File: nodejs/src/controllers/AuthController.ts
import { Request, Response } from 'express';
import jwt from 'jsonwebtoken';
import { LoginDto, LoginResponseDto } from '../dto/LoginDto';

export class AuthController {
  static async login(req: Request, res: Response): Promise<void> {
    try {
      const { username, password }: LoginDto = req.body;

      // Usuarios demo (en producción esto vendría de una base de datos)
      const users = [
        { username: 'admin', password: 'admin123', role: 'Admin' },
        { username: 'consultor', password: 'consultor123', role: 'Consultor' }
      ];

      const user = users.find(u => u.username === username && u.password === password);

      if (!user) {
        res.status(401).json({ message: 'Credenciales inválidas' });
        return;
      }

      const token = jwt.sign(
        { 
          username: user.username, 
          role: user.role 
        },
        process.env.JWT_SECRET!,
        { 
          expiresIn: process.env.JWT_EXPIRES_IN || '24h',
          issuer: process.env.JWT_ISSUER || 'PersonasABM',
          audience: process.env.JWT_AUDIENCE || 'PersonasABMUsers'
        }
      );

      const response: LoginResponseDto = {
        token,
        username: user.username,
        role: user.role,
        message: 'Login exitoso'
      };

      res.json(response);
    } catch (error) {
      console.error('Error en login:', error);
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }
}

File: dotnet/PersonasABM.API/Controllers/AuthController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;

namespace PersonasABM.API.Controllers;

[ApiController]
[Route("api/[controller]")]
public class AuthController : ControllerBase
{
    private readonly IConfiguration _configuration;

    public AuthController(IConfiguration configuration)
    {
        _configuration = configuration;
    }

    [HttpPost("login")]
    [AllowAnonymous]
    public IActionResult Login([FromBody] LoginRequest request)
    {
        // Usuarios demo (en producción esto vendría de una base de datos)
        var users = new[]
        {
            new { Username = "admin", Password = "admin123", Role = "Admin" },
            new { Username = "consultor", Password = "consultor123", Role = "Consultor" }
        };

        var user = users.FirstOrDefault(u => u.Username == request.Username && u.Password == request.Password);
        
        if (user == null)
        {
            return Unauthorized(new { message = "Credenciales inválidas" });
        }

        var token = GenerateJwtToken(user.Username, user.Role);
        
        return Ok(new
        {
            token,
            username = user.Username,
            role = user.Role,
            message = "Login exitoso"
        });
    }

    private string GenerateJwtToken(string username, string role)
    {
        var jwtSettings = _configuration.GetSection("JwtSettings");
        var secretKey = jwtSettings["SecretKey"] ?? "MiClaveSecretaSuperSeguraParaJWT2024!";
        var issuer = jwtSettings["Issuer"] ?? "PersonasABM";
        var audience = jwtSettings["Audience"] ?? "PersonasABMUsers";
        var expirationHours = int.Parse(jwtSettings["ExpirationHours"] ?? "24");

        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(secretKey));
        var credentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);

        var claims = new[]
        {
            new Claim(ClaimTypes.Name, username),
            new Claim(ClaimTypes.Role, role),
            new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
            new Claim(JwtRegisteredClaimNames.Iat, new DateTimeOffset(DateTime.UtcNow).ToUnixTimeSeconds().ToString(), ClaimValueTypes.Integer64)
        };

        var token = new JwtSecurityToken(
            issuer: issuer,
            audience: audience,
            claims: claims,
            expires: DateTime.UtcNow.AddHours(expirationHours),
            signingCredentials: credentials
        );

        return new JwtSecurityTokenHandler().WriteToken(token);
    }
}

public class LoginRequest
{
    public string Username { get; set; } = string.Empty;
    public string Password { get; set; } = string.Empty;
}
